packages/vite/src/node/server/index.ts

```diff
@@ -501,12 +497,10 @@ export async function createServer(
   // serve custom content instead of index.html.
   postHooks.forEach((fn) => fn && fn())
 
-  if (config.spa && !isMiddlewareMode) {
+  if (config.appType === 'spa' || config.appType === 'mpa') {
     // transform index.html
     middlewares.use(indexHtmlMiddleware(server))
-  }
 
-  if (!isMiddlewareMode) {
```

Comment:
Should we not remove this, so we let the consumer handle un-handled requests instead? I think that was the goal of middlewareMode

Reply:
They can do that by using `appType: 'custom'`, no? The idea of removing this line is for `middlewareMode` only define if Vite controls the server or not.

Reply:
This if condition is conditionally applying a 404 middleware though, which wouldn't call `next()`. 

If we have:
```js
  const vite = await createViteServer({
    server: { middlewareMode: true },
    appType: 'custom'
  })
  app.use(vite.middlewares)
  app.use('*', async (req, res) => {
    // This will never get called because Vite captures everything
  })
```

The middlewares after Vite would never get used.

Reply:
If you set `appType: 'custom'`, this middleware is not going to be added so that snippet of code will work, no?

Reply:
The idea was that `appType: 'spa'` + `midlewareMode: true` would give you the same as `midlewareMode: 'html'`. Maybe we should remove that comment and document this better stating that middlewares should be added before vite ones (or even in the middle as you can modify the array) for 'spa' or 'mpa'? When selecting `appType: 'spa'` I think users are buying into the way Vite middlewares are handled including this 404 that if we don't add then the user will need to add by hand, no?

Reply:
Ah ok I think you're right. And checking the v2 code, we indeed handle 404 for `middlewareMode: 'html'`

https://github.com/vitejs/vite/blob/86a55d3cc0668eca79a55f5cf8b6034b9e3bf835/packages/vite/src/node/server/index.ts#L564-L573

Thanks for taking the time to clarify this. I think updating the docs make sense, I thought there would be more going on for `spa` and `mpa` too when they'd never be reached.

Reply:
Done in [chore: update docs](https://github.com/vitejs/vite/pull/8452/commits/c061a02c1afe3c0357be5e1ef0dcd80d5b59db3e)